nemo/collections/nlp/models/nlp_model.py

```diff
@@ -0,0 +1,123 @@
+                app_state.model_parallel_rank = torch.distributed.get_rank(group=app_state.model_parallel_group)
+                app_state.data_parallel_rank = torch.distributed.get_rank(group=app_state.data_parallel_group)
+                logging.info(f'mp_rank: {app_state.model_parallel_rank}')
+                logging.info(f'dp_rank: {app_state.data_parallel_rank}')
+
+    def configure_ddp(self, model: LightningModule, device_ids: List[int]) -> DistributedDataParallel:
+        """ Override LightningModule ddp if using model parallel.
+
+        Args:
+            model (LightningModule): the LightningModule currently being optimized
+            device_ids (List[int]): the list of GPU ids.
+
+        Returns:
+            DistributedDataParallel: DDP wrapped model
+        """
+
+        app_state = AppState()
+
+        if app_state.model_parallel_size is not None:
+            logging.info("Configuring DDP for model parallelism.")
+            logging.info(f"data_parallel_group: {app_state.data_parallel_group}")
+            # with model parallelism, multiple GPUs form a large "logical GPU"
+            # this means that data parallel groups span multiple GPUs
+            # and are non-trivial
+
+            model = LightningDistributedDataParallel(
+                model, device_ids, output_device=device_ids[0], process_group=app_state.data_parallel_group
+            )
+            return model
+
+        else:
+            logging.info("Did not detect model parallel using LightningModule.configure_ddp")
+            return LightningModule.configure_ddp(self, model, device_ids)
+
+    def setup(self, stage: str) -> None:
+        """ PTL hook that is called after DDP is initialized.
+            Called at the beginning of fit and test. 
+
+        Args:
+            stage (str): either 'fit' or 'test'
+        """
+
+        if stage == 'fit':
+
+            app_state = AppState()
+
+            if app_state.model_parallel_size is not None:
+                if isinstance(self.bert_model, MegatronBertEncoder):
+                    logging.info(f"restoring model parallel checkpoint: {self.bert_model._restore_path}")
+                    # model parallel checkpoints need to be restored after torch.distributed is initialized
+                    self.bert_model.restore_weights(self.bert_model._restore_path)
+
+                    logging.info("replacing sampler with model parallel sampler")
+                    mp_sampler = torch.utils.data.distributed.DistributedSampler(
+                        self._train_dl.dataset,
+                        num_replicas=app_state.model_parallel_size,
+                        rank=app_state.data_parallel_rank,
+                    )
+                    mp_dl = self._trainer.replace_sampler(self._train_dl, mp_sampler)
+                    self._train_dl = mp_dl
+                else:
+                    logging.warning(f'The BERT encoder: {self.bert_model} does not support model parallelism.')
```

Comment:
shouldn't we raise NotImplemented exception here?

Reply:
Hard crash would be better here to avoid user confusion.